## Задача 3-1. 
Даны значения зарплат из выборки выпускников: 100, 80, 75, 77, 89, 33, 45, 25, 65, 17, 30, 24, 57, 55, 70, 75, 65, 84, 90, 150. Посчитать (желательно без использования статистических методов наподобие std, var, mean) среднее арифметическое, среднее квадратичное отклонение, смещенную и несмещенную оценки дисперсий для данной выборки.
Можно затем посчитать те же значения с использованием статистических функций с целью самопроверки.

### Решение:
[1] Cреднее арифметическое зарплат выпускников: $M(X)={\frac{1}{n}}{\displaystyle\sum_{i=1}^{n} x_i}$

[2] Среднее квадратичное отклонение зарплат выпускников: $\sigma = \sqrt{\frac{\displaystyle\sum\limits_{i=1}^{n} (x_i - \overline{x})^2}{n}}$

[3] Смещенная оценка дисперсии зарплат выпускников: $\sigma^2_{biased} = \frac{\displaystyle\sum\limits_{i=1}^{n} (x_i - \overline{x})^2}{n}$

[4] Несмещенная оценка дисперсии зарплат выпускников: $\sigma^2_{unbiased} = \frac{\displaystyle\sum\limits_{i=1}^{n} (x_i - \overline{x})^2}{n - 1}$

In [7]:
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt
import warnings
warnings.filterwarnings('ignore')

In [4]:
def mean_salary(salary):
    if len(salary) > 0:
        mean_salary_1 = salary.sum() / len(salary)
        mean_salary_2 = salary.mean()
        print(f'Среднее арифметическое равно {mean_salary_1} (проверка {mean_salary_2})')

def std_salary(salary):
    if len(salary) > 0:
        std_salary_1 = np.sqrt(((salary - salary.mean()) ** 2).sum() / len(salary))
        std_salary_2 = salary.std(ddof=0)
        print(f'Среднее квадратичное отклонение {std_salary_1} (проверка {std_salary_2})')

def variance_salary1(salary):
    if len(salary) > 0:
        variance_salary_1 = ((salary - salary.mean()) ** 2).sum() / len(salary)
        variance_salary_2 = salary.var(ddof=0)
        print(f'Cмещенная оценка дисперсии  {variance_salary_1} (проверка {variance_salary_2})')

def variance_salary2(salary):
    if len(salary) > 0:
        variance_salary2_1 = ((salary - salary.mean()) ** 2).sum() / (len(salary) - 1)
        variance_salary2_2 = salary.var(ddof=1)
        print(f'Несмещенная оценка дисперсии  {variance_salary2_1} (проверка {variance_salary2_2})')

In [6]:
salary = np.array([100, 80, 75, 77, 89, 33, 45, 25, 65, 17, 30, 24, 57, 55, 70, 75, 65, 84, 90, 150])
mean_salary(salary)
std_salary(salary)
variance_salary1(salary)
variance_salary2(salary)

Среднее арифметическое равно 65.3 (проверка 65.3)
Среднее квадратичное отклонение 30.823854398825596 (проверка 30.823854398825596)
Cмещенная оценка дисперсии  950.11 (проверка 950.11)
Несмещенная оценка дисперсии  1000.1157894736842 (проверка 1000.1157894736842)


#### Ответ 3-1: cреднее арифметическое зарплат выпускников: $M = 65,3$; cреднее квадратичное отклонение зарплат выпускников $\sigma = 30,82385$; cмещенная оценка дисперсии зарплат выпускников  $\sigma^2_{biased} = 950,11$; несмещенная оценка дисперсии зарплат выпускников $\sigma^2_{unbiased} = 1000,115789$

## Задача 3-2. 
В первом ящике находится 8 мячей, из которых 5 - белые. Во втором ящике - 12 мячей, из которых 5 белых. Из первого ящика вытаскивают случайным образом два мяча, из второго - 4. Какова вероятность того, что 3 мяча белые?

### Решение:
[1] Вероятнось вынуть 0 белых мячей из 1-го ящика и 3 белых из 2-го ящика: $P_1=\left({\frac{{C_{5}^0}*{C_{3}^2}}{C_{8}^2}}\right)*\left({\frac{{C_{5}^3}*{C_{7}^1}}{C_{12}^4}}\right)=\left({\frac{{\frac{5!}{0!(5-0)!}}*{\frac{3!}{2!(3-2)!}}}{\frac{8!}{2!(8-2)!}}}\right)*\left({\frac{{\frac{5!}{3!(5-3)!}}*{\frac{7!}{1!(7-1)!}}}{\frac{12!}{4!(12-4)!}}}\right)$

[2] Вероятнось вынуть 1 белый мяч из 1-го ящика и 2 белых из 2-го ящика: $P_2=\left({\frac{{C_{5}^1}*{C_{3}^1}}{C_{8}^2}}\right)*\left({\frac{{C_{5}^2}*{C_{7}^2}}{C_{12}^4}}\right)=\left({\frac{{\frac{5!}{1!(5-1)!}}*{\frac{3!}{1!(3-1)!}}}{\frac{8!}{2!(8-2)!}}}\right)*\left({\frac{{\frac{5!}{2!(5-2)!}}*{\frac{7!}{2!(7-2)!}}}{\frac{12!}{4!(12-4)!}}}\right)$

[3] Вероятнось вынуть 2 белых мяча из 1-го ящика и 1 белый мяч из 2-го ящика: $P_3=\left({\frac{{C_{5}^2}*{C_{3}^0}}{C_{8}^2}}\right)*\left({\frac{{C_{5}^1}*{C_{7}^3}}{C_{12}^4}}\right)=\left({\frac{{\frac{5!}{2!(5-2)!}}*{\frac{3!}{0!(3-0)!}}}{\frac{8!}{2!(8-2)!}}}\right)*\left({\frac{{\frac{5!}{1!(5-1)!}}*{\frac{7!}{3!(7-3)!}}}{\frac{12!}{4!(12-4)!}}}\right)$

[4] Вероятность того, что 3 мяча белые: $P=P_1+P_2+P_3$

In [10]:
from scipy.special import comb
p1 = ((comb(5, 0, exact = False, repetition=False) * (comb(3, 2, exact = False, repetition=False))) / comb(8, 2, exact = False, repetition=False)) * (((comb(5, 3, exact = False, repetition=False)) * (comb(7, 1, exact = False, repetition=False))) / comb(12, 4, exact = False, repetition=False))
p2 = ((comb(5, 1, exact = False, repetition=False) * (comb(3, 1, exact = False, repetition=False))) / comb(8, 2, exact = False, repetition=False)) * (((comb(5, 2, exact = False, repetition=False)) * (comb(7, 2, exact = False, repetition=False))) / comb(12, 4, exact = False, repetition=False))
p3 = ((comb(5, 2, exact = False, repetition=False) * (comb(3, 0, exact = False, repetition=False))) / comb(8, 2, exact = False, repetition=False)) * (((comb(5, 1, exact = False, repetition=False)) * (comb(7, 3, exact = False, repetition=False))) / comb(12, 4, exact = False, repetition=False))
print(f'Вероятнось вынуть 0 белых мячей из 1-го ящика и 3 белых из 2-го ящика = {round(p1, 7)}\n'
      f'Вероятнось вынуть 1 белый мяч из 1-го ящика и 2 белых из 2-го ящика = {round(p2, 7)}\n'
      f'Вероятнось вынуть 2 белых мяча из 1-го ящика и 1 белый из 2-го ящика = {round(p3, 7)}\n'
      f'Вероятность того, что ровно два мяча белые = {round((p1 + p2 + p3), 7)}')

Вероятнось вынуть 0 белых мячей из 1-го ящика и 3 белых из 2-го ящика = 0.0151515
Вероятнось вынуть 1 белый мяч из 1-го ящика и 2 белых из 2-го ящика = 0.2272727
Вероятнось вынуть 2 белых мяча из 1-го ящика и 1 белый из 2-го ящика = 0.1262626
Вероятность того, что ровно два мяча белые = 0.3686869


#### Ответ 3-2: вероятность того, что три мяча белые составляет 36,86(86)%

## Задача 3-3.
На соревновании по биатлону один из трех спортсменов стреляет и попадает в мишень. Вероятность попадания для первого спортсмена равна 0.9, для второго — 0.8, для третьего — 0.6.

Найти вероятность того, что выстрел произведен:   a). первым спортсменом    б). вторым спортсменом    в). третьим спортсменом.

### Решение:
#### Формула Байеса
Чтобы определить вероятность события $B$ при условии, что событие $A$ уже произошло, используют формулу Байеса:
$$P(B\:|\:A) = \frac{P(B) \cdot P(A\:|\:B)}{P(A)}$$
В данной формуле вероятность $P(B\;|\;A)$ называется апостериорной, то есть определенной после того, как произошло событие $A$. 

А вероятность $P(B)$ — априорной, определенной до испытания.
Прежде чем узнали результат выстрела, мы располагали только априорной вероятностью того, что выстрел был произведен, и она равнялась 1/3 — если учесть, что у всех спортсменов были равные возможности выстрелить.

Зная результат выстрела, мы можем заново оценить вероятности для каждого спортсмена — ведь точность у них разная. Эти вероятности будут апостериорными, так как они получены после поступления информации о том, что выстрел был удачным.

Событием $A$ будем считать попадание в мишень, а событиями $B_1, B_2$ и $B_3$ — что выстрел совершил первый, второй или третий спортсмен:

$P(A\:|\:B_1) = 0.9$

$P(A\:|\:B_2) = 0.8$

$P(A\:|\:B_3) = 0.6$

a). Вероятность того, что стрелял первый спортсмен — при условии, что выстрел был удачным, — будем находить по формуле:
$$P(B_1\:|\:A) = \frac{P(B_1) \cdot P(A\:|\:B_1)}{P(A)}=\frac{\frac{1}{3} \cdot 0.9}{\frac{1}{3} \cdot 0.9 + \frac{1}{3} \cdot 0.8 + \frac{1}{3} \cdot 0.6}$$

б). Вероятность того, что стрелял второй спортсмен — при условии, что выстрел был удачным, — будем находить по формуле:
$$P(B_2\:|\:A) = \frac{P(B_2) \cdot P(A\:|\:B_2)}{P(A)}=\frac{\frac{1}{3} \cdot 0.8}{\frac{1}{3} \cdot 0.9 + \frac{1}{3} \cdot 0.8 + \frac{1}{3} \cdot 0.6}$$

в). Вероятность того, что стрелял третий спортсмен — при условии, что выстрел был удачным, — будем находить по формуле:
$$P(B_3\:|\:A) = \frac{P(B_3) \cdot P(A\:|\:B_3)}{P(A)}=\frac{\frac{1}{3} \cdot 0.6}{\frac{1}{3} \cdot 0.9 + \frac{1}{3} \cdot 0.8 + \frac{1}{3} \cdot 0.6}$$

In [18]:
p_1 = 0.9
p_2 = 0.8
p_3 = 0.6
p_A = 1/3 * p_1 + 1/3 * p_2 + 1/3 * p_3
p_B1 = (1/3 * p_1) / p_A
p_B2 = (1/3 * p_2) / p_A
p_B3 = (1/3 * p_3) / p_A
print(f'Полная вероятность попадания в мишень всеми спортсменами: 𝑃(𝐴) = {round(p_A, 7)}\n'
      f'Вероятность того, что стрелял первый спортсмен: 𝑃(𝐵1|𝐴) = {round(p_B1, 7)}\n'
      f'Вероятность того, что стрелял второй спортсмен: 𝑃(𝐵2|𝐴) = {round(p_B2, 7)}\n'
      f'Вероятность того, что стрелял третий спортсмен: 𝑃(𝐵3|𝐴) = {round(p_B3, 7)}\n'
      f'Проверка верности расчетов: 𝑃(𝐵1|𝐴) + 𝑃(𝐵2|𝐴) + 𝑃(𝐵3|𝐴) = {round((p_B1 + p_B2 + p_B3), 7)}')

Полная вероятность попадания в мишень всеми спортсменами: 𝑃(𝐴) = 0.7666667
Вероятность того, что стрелял первый спортсмен: 𝑃(𝐵1|𝐴) = 0.3913043
Вероятность того, что стрелял второй спортсмен: 𝑃(𝐵2|𝐴) = 0.3478261
Вероятность того, что стрелял третий спортсмен: 𝑃(𝐵3|𝐴) = 0.2608696
Проверка верности расчетов: 𝑃(𝐵1|𝐴) + 𝑃(𝐵2|𝐴) + 𝑃(𝐵3|𝐴) = 1.0


#### Ответ 3-3: вероятность того, что выстрел произведен: 
#### a). первым спортсменом $P(B_1\:|\:A) = 39,13\%$
#### б). вторым спортсменом $P(B_2\:|\:A) = 34,78\%$
#### в). третьим спортсменом $P(B_3\:|\:A) = 26,09\%$

## Задача 3-4.
В университет на факультеты A и B поступило равное количество студентов, а на факультет C студентов поступило столько же, сколько на A и B вместе. Вероятность того, что студент факультета A сдаст первую сессию, равна 0.8. Для студента факультета B эта вероятность равна 0.7, а для студента факультета C - 0.9. Студент сдал первую сессию.
Какова вероятность, что он учится: a). на факультете A б). на факультете B в). на факультете C?

### Решение:
#### Формула Байеса
Чтобы определить вероятность события $B$ при условии, что событие $A$ уже произошло, используют формулу Байеса:
$$P(B\:|\:A) = \frac{P(B) \cdot P(A\:|\:B)}{P(A)}$$
[1] A и B равны и являются половиной C, то очевидно, что вероятность случайного выбора:

Вероятность, что студент учится на факультете А: $P(F_A) = 0.25$

Вероятность, что студент учится на факультете B: $P(F_B) = 0.25$

Вероятность, что студент учится на факультете C: $P(F_C) = 0.5$

[2] Возьмем из условия вероятности сдачи каждого студента:

$P(F_A\:|\:A) = 0.8$

$P(F_B\:|\:A) = 0.7$

$P(F_C\:|\:A) = 0.9$

[3] Найдем полную вероятность, что сдатут все студенты:

$P(A) = P(F_A) \cdot P(F_A\:|\:A) + P(F_B) \cdot P(F_B\:|\:A) + P(F_C) \cdot P(F_C\:|\:A) $

[4] Найдем вероятности согласно формуле Байеса:

Вероятность того, что студент учится на факультете A: $𝑃(F_A) = \frac{P(F_A) \cdot P(F_A\:|\:A)}{P(A)}$

Вероятность того, что студент учится на факультете B: $𝑃(F_B) = \frac{P(F_B) \cdot P(F_B\:|\:A)}{P(A)}$

Вероятность того, что студент учится на факультете C: $𝑃(F_C) = \frac{P(F_C) \cdot P(F_C\:|\:A)}{P(A)}$

In [28]:
p_1 = 0.8
p_2 = 0.7
p_3 = 0.9
p_A = (0.25 * p_1) + (0.25 * p_2) + (0.5 * p_3)
p_F_A = (0.25 * p_1) / p_A
p_F_B = (0.25 * p_2) / p_A
p_F_C = (0.5 * p_3) / p_A
print(f'Полная вероятность, что сдатут все студенты: 𝑃(𝐴) = {round(p_A, 7)}\n'
      f'Вероятность того, что студент учится на факультете A: 𝑃(𝐹𝐴) = {round(p_F_A, 7)}\n'
      f'Вероятность того, что студент учится на факультете B: 𝑃(𝐹𝐵) = {round(p_F_B, 7)}\n'
      f'Вероятность того, что студент учится на факультете C: 𝑃(𝐹𝐶) = {round(p_F_C, 7)}\n'
      f'Проверка верности расчетов: 𝑃(FA) + 𝑃(F𝐵) + 𝑃(FC) = {round((p_F_A + p_F_B + p_F_C), 7)}')

Полная вероятность, что сдатут все студенты: 𝑃(𝐴) = 0.825
Вероятность того, что студент учится на факультете A: 𝑃(𝐹𝐴) = 0.2424242
Вероятность того, что студент учится на факультете B: 𝑃(𝐹𝐵) = 0.2121212
Вероятность того, что студент учится на факультете C: 𝑃(𝐹𝐶) = 0.5454545
Проверка верности расчетов: 𝑃(FA) + 𝑃(F𝐵) + 𝑃(FC) = 1.0


#### Ответ 3-4: вероятность того, что студент учится: 
#### a). на факультете A: $𝑃(𝐹_𝐴) = 24,24(24)\%$
#### б). на факультете B: $P(F_B) = 21,21(21)\%$
#### в). на факультете C: $P(F_C) = 54,54(54)\%$

## Задача 3-5.
Устройство состоит из трех деталей. Для первой детали вероятность выйти из строя в первый месяц равна 0.1, для второй - 0.2, для третьей - 0.25. Какова вероятность того, что в первый месяц выйдут из строя:
а). все детали
б). только две детали
в). хотя бы одна деталь
г). от одной до двух деталей?

### Решение:

In [6]:
p_1 = 0.1
p_2 = 0.2
p_3 = 0.25
p_all = p_1 * p_2 * p_3
p_only_2 = ((1-p_1)*p_2*p_3)+(p_1*(1-p_2)*p_3)+(p_1*p_2*(1-p_3))
p_at_least_1 = 1-((1-p_1)*(1-p_2)*(1-p_3))
print(f'Вероятность, что из строя выйдут все детали: 𝑃(all) = {p_all}\n'
      f'Вероятность, что из строя выйдут только две детали: 𝑃(only 2) = {p_only_2}\n'
      f'Вероятность того, из строя выйдет хотя бы одна деталь: 𝑃(at least 1) = {p_at_least_1}\n'
      f'Вероятность того, из строя выйдут от одной до двух деталей: 𝑃(1 till 2) = {p_at_least_1 - p_all}\n')

Вероятность, что из строя выйдут все детали: 𝑃(all) = 0.005000000000000001
Вероятность, что из строя выйдут только две детали: 𝑃(only 2) = 0.08
Вероятность того, из строя выйдет хотя бы одна деталь: 𝑃(at least 1) = 0.45999999999999996
Вероятность того, из строя выйдут от одной до двух деталей: 𝑃(1 till 2) = 0.45499999999999996



#### Ответ 3-5: вероятность того, что в первый месяц выйдут из строя: 
#### a). все детали: $𝑃_{all} = 0,5\%$
#### б). только две детали: $P_{only~2} = 8\%$
#### в). хотя бы одна деталь: $P_{at~least~1} = 46\%$
#### г). от одной до двух деталей: $P_{1~till~2} = 45,5\%$